### PR TITLE
[DAGCombiner] Inverse transform `(select c, (and X, 1), 0)` -> `(and (zext c), X)`

### PR DESCRIPTION
**Pull request by goldsteinn**
- [X86][AArch64][RISCV] Add tests for combining `(select c, (and X, 1), 0)` -> `(and (zext c), X)`; NFC
- [DAGCombiner] Combine `(select c, (and X, 1), 0)` -> `(and (zext c), X)`

### CODE REVIEW

File: llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp [hunk 1]
```diff
@@ -6294,6 +6294,37 @@ static SDValue foldAndOrOfSETCC(SDNode *LogicOp, SelectionDAG &DAG) {
...
 }
 
+// Combine `(select c, (X & 1), 0)` -> `(and (zext c), X)`.
+// We canonicalize to the `select` form in the middle end, but the `and` form
+// gets better codegen and all tested targets (arm, x86, riscv)
+static SDValue combineSelectAsExtAnd(SDValue Cond, SDValue T, SDValue F,
+                                     const SDLoc &DL, SelectionDAG &DAG) {
+  const TargetLowering &TLI = DAG.getTargetLoweringInfo();
+  auto *FC = dyn_cast<ConstantSDNode>(F);
+  if (FC == nullptr || !FC->isZero())
```
Comment (topperc) [inline llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:6304]: !isNullConstant(F)?

File: llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp [hunk 2]
```diff
@@ -6294,6 +6294,37 @@ static SDValue foldAndOrOfSETCC(SDNode *LogicOp, SelectionDAG &DAG) {
...
+  ISD::NodeType BoolExtOpc =
+      TLI.getExtendForContent(TLI.getBooleanContents(CondVT));
+  if (BoolExtOpc != ISD::ZERO_EXTEND)
+    return SDValue();
+
+  if (T.getOpcode() != ISD::AND)
+    return SDValue();
+
+  auto *TC1 = dyn_cast<ConstantSDNode>(T.getOperand(1));
+  if (TC1 == nullptr || !TC1->isOne())
```
Comment (topperc) [inline llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:6317]: !isOneConstant(T)?

File: llvm/test/CodeGen/X86/select-to-and-zext.ll [hunk 3]
```diff
@@ -0,0 +1,189 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py
+; RUN: llc < %s -mtriple=i686-pc-linux-gnu | FileCheck %s --check-prefix=X86
+; RUN: llc < %s -mtriple=x86_64-pc-linux-gnu | FileCheck %s --check-prefix=X64
+
+define i32 @from_cmpeq(i32 %xx, i32 %y) {
```
Comment (RKSimon) [inline llvm/test/CodeGen/X86/select-to-and-zext.ll:5]: Please can you regenerate this so that it doesn't have the encoding comments

**Comment by goldsteinn** [comment]
> Not sure whats going on with these stacked PRs - but this needs an actual patch title (and not the branch name)

Sorry, growing pains with the github workflow.

**Comment by goldsteinn** [comment]
ping.

**Review by RKSimon — APPROVED** [review]
LGTM - cheers

**Comment by goldsteinn** [comment]
Pushed.

**Comment by nikic** [comment]
@goldsteinn If you manually push a change, you need to do something like this
```
git checkout foo
git fetch upstream
git rebase upstream/main

git push -f fork foo
git push upstream foo:main
```
to keep the association between the PR and the commits (the salient part being the last two lines).

If you don't do that, please edit the commit messages to include something like "Closes #xyz" to create a back-link to the PR.

**Pull request closed without merging**